File: packages/sdk-py/pylva/core/context.py
"""Propagation context via contextvars — D25.

``track(customer_id, ...)`` nests cleanly for both sync and async callers.
The context is async-safe without the caller using ``copy_context`` — the
common Python async pattern.
"""

from __future__ import annotations

import inspect
import re
import uuid
from collections.abc import Awaitable, Callable, Iterator
from contextlib import contextmanager
from contextvars import ContextVar, Token
from dataclasses import dataclass
from typing import TypeVar


@dataclass(frozen=True)
class TrackContext:
    customer_id: str
    trace_id: str
    span_id: str
    parent_span_id: str | None
    step_name: str | None
    framework: str
    run_id: str
    parent_run_id: str | None


_context: ContextVar[TrackContext | None] = ContextVar("pylva_context", default=None)

T = TypeVar("T")


def current_context() -> TrackContext | None:
    return _context.get()


# Matches the backend's per-event customer_id schema. track() warns (never
# raises — R1) so builders learn at dev time instead of discovering that the
# backend rejected the events: rejected traffic is unbilled AND budget rules
# never apply to it.
_CUSTOMER_ID_RE = re.compile(r"^[a-zA-Z0-9_-]{1,255}$")
_warned_invalid_customer_ids: set[str] = set()


def _warn_invalid_customer_id(customer_id: object) -> None:
    key = str(customer_id)[:64]
    if key in _warned_invalid_customer_ids or len(_warned_invalid_customer_ids) > 1000:
        return
    _warned_invalid_customer_ids.add(key)
    print(
        f"[pylva] track(): customer_id {key!r} does not match ^[a-zA-Z0-9_-]{{1,255}}$ — "
        "the backend will reject its events, so this traffic is unbilled and budget rules "
        "will not apply to it. Use a stable alphanumeric/_/- id.",
        flush=True,
    )


@contextmanager
def track_context(
    customer_id: str,
    step: str | None = None,
    framework: str | None = None,
) -> Iterator[TrackContext]:
    if not isinstance(customer_id, str) or not _CUSTOMER_ID_RE.match(customer_id):
        _warn_invalid_customer_id(customer_id)
    parent = _context.get()
    ctx = TrackContext(
        customer_id=customer_id,
        trace_id=parent.trace_id if parent else str(uuid.uuid4()),
        span_id=str(uuid.uuid4()),
        parent_span_id=parent.span_id if parent else None,
        step_name=step or (parent.step_name if parent else None),
        framework=framework or (parent.framework if parent else "none"),
        run_id=str(uuid.uuid4()),
        parent_run_id=parent.run_id if parent else None,
    )
    token: Token[TrackContext | None] = _context.set(ctx)
    try:
        yield ctx
    finally:
        _context.reset(token)


def track(
    customer_id: str,
    fn: Callable[[], T | Awaitable[T]],
    *,
    step: str | None = None,
    framework: str | None = None,
) -> T | Awaitable[T]:
    """Run ``fn`` inside a track context. ``fn`` may be sync or async.

    Usage:
        result = track("cust_123", lambda: do_work())
        # async:
        result = await track("cust_123", async_do_work)
    """
    with track_context(customer_id, step, framework) as ctx:
        result = fn()
        if not inspect.isawaitable(result):
            return result
    # Async fn: ``fn()`` only created the coroutine — its body runs when the
    # caller awaits, *after* this function has returned and the ``with`` block
    # reset the contextvar. Re-enter the context for the awaited execution so
    # LLM calls inside attribute to ``customer_id`` instead of "anonymous"
    # (and customer-scoped budget rules still match). Mirrors TS
    # ``AsyncLocalStorage.run()``, which keeps context across awaits.
    return _await_in_context(ctx, result)


async def _await_in_context(ctx: TrackContext, awaitable: Awaitable[T]) -> T:
    token = _context.set(ctx)
    try:
        return await awaitable
    finally:
        _context.reset(token)


def _run_in_context(ctx: TrackContext, fn: Callable[[], T]) -> T:
    token = _context.set(ctx)
    try:
        return fn()
    finally:
        _context.reset(token)

File: packages/sdk-ts/src/core/context.ts
// Propagation context via AsyncLocalStorage.
// pylva.track(customerId, opts?, cb) nests: child spans inherit trace_id
// and use the parent's span_id as parent_span_id.

import { AsyncLocalStorage } from 'node:async_hooks';
import { randomUUID } from 'node:crypto';
import { Framework, type Framework as FrameworkType } from '@pylva/shared/telemetry-values';

export interface TrackContext {
  customer_id: string;
  trace_id: string;
  span_id: string;
  parent_span_id: string | null;
  step_name: string | null;
  framework?: FrameworkType;
  run_id: string;
  parent_run_id: string | null;
}

// Published entrypoints import this one physical canonical CJS module. Its
// module cache owns the ALS; no discoverable global rendezvous is involved.
const als = new AsyncLocalStorage<TrackContext>();

export function currentContext(): TrackContext | undefined {
  return als.getStore();
}

// Matches the backend's per-event customer_id schema. track() warns (never
// throws — R1) so builders learn at dev time instead of discovering that the
// backend rejected the events: rejected traffic is unbilled AND budget rules
// never apply to it.
const CUSTOMER_ID_RE = /^[a-zA-Z0-9_-]{1,255}$/;
const warnedInvalidCustomerIds = new Set<string>();

function warnInvalidCustomerId(customerId: unknown): void {
  const key = String(customerId).slice(0, 64);
  if (warnedInvalidCustomerIds.has(key) || warnedInvalidCustomerIds.size > 1000) return;
  warnedInvalidCustomerIds.add(key);
  console.warn(
    `[pylva] track(): customer_id ${JSON.stringify(key)} does not match ^[a-zA-Z0-9_-]{1,255}$ — ` +
      'the backend will reject its events, so this traffic is unbilled and budget rules will not ' +
      'apply to it. Use a stable alphanumeric/_/- id.',
  );
}

export interface TrackOptions {
  step?: string;
  framework?: FrameworkType;
}

type TrackCallback<T> = () => Promise<T> | T;

export function track<T>(customerId: string, cb: TrackCallback<T>): Promise<T> | T;
export function track<T>(
  customerId: string,
  opts: TrackOptions,
  cb: TrackCallback<T>,
): Promise<T> | T;
export function track<T>(
  customerId: string,
  optsOrCb: TrackOptions | TrackCallback<T>,
  cb?: TrackCallback<T>,
): Promise<T> | T {
  let options: TrackOptions = {};
  let callback: TrackCallback<T>;
  if (typeof optsOrCb === 'function') {
    callback = optsOrCb;
  } else {
    options = optsOrCb;
    if (!cb) {
      throw new Error('[pylva] track(customerId, opts, cb): cb is required when opts is provided');
    }
    callback = cb;
  }

  if (typeof customerId !== 'string' || !CUSTOMER_ID_RE.test(customerId)) {
    warnInvalidCustomerId(customerId);
  }

  const parent = als.getStore();
  const ctx: TrackContext = {
    customer_id: customerId,
    trace_id: parent?.trace_id ?? randomUUID(),
    span_id: randomUUID(),
    parent_span_id: parent?.span_id ?? null,
    step_name: options.step ?? parent?.step_name ?? null,
    framework: options.framework ?? parent?.framework ?? Framework.NONE,
    run_id: randomUUID(),
    parent_run_id: parent?.run_id ?? null,
  };

  return als.run(ctx, callback);
}

// Test-only helper; not exported from index.
export function _runInContext<T>(ctx: TrackContext, cb: TrackCallback<T>): Promise<T> | T {
  return als.run(ctx, cb);
}
